### ISSUE TITLE
more generic parser

### ISSUE THREAD

**Issue by ulno** [Daenara/rhasspy_weather#3]
Would it be possible to abstract out a more generic parser that wouldn't have the slot mechanic so hard coded into it.

So far, I based https://github.com/ulno/cli_weather/blob/master/cli_parser/__init__.py on https://github.com/Daenara/rhasspy_weather/blob/master/parser/rhasspy_intent.py, though still replicating a lot of functionality. Can we define some kind of parser interface, allowing the merge of the common functionality into one module?

**Comment by Daenara** [comment]
That parser is literally only made to parse rhasspy_intents. My idea was that all parsers throw our a WeatherRequest object and the rest of the script works with that. I personally see no way to extract anything out from there for a more general use because it basically just takes the value from a slot and figures out how exactly it has to write that into a WeatherRequest.

I could start writing functions that takes values of a certain type and then return part of the WeatherRequest but while that may for the way I use it and the way you use it it would not be universal for every usage ppl might come up and then they would need to format their input into a format that fits the functions that output into the format of the WeatherRequest and that like a situation to avoid.

If you see a way to avoid that problem and still generalize stuff I am open for discussion but the WeatherRequest right now is the general interface that all parsers should use, the rest is parser intern.

**Comment by ulno** [comment]
I suggest splitting up the big parsing with all the ifs into subfunctions for each respective slot, interpreting different slots. Then these subfunctions can be called also from other parsers. 

For example:
```python
        def parse_time(self, time_string):
             time_string=time_string.lower()
             log.debug("it was a time specified: {}".format(time_string})
            self.new_request.grain = Grain.HOUR

            named_times_lowercase = [x.lower() for x in config.locale.named_times]
            named_times_synonyms_lowercase = [x.lower() for x in config.locale.named_times_synonyms]
            named_times_combined = named_times_lowercase + named_times_synonyms_lowercase
            # was something like midday specified (listed in locale.named_times or in locale.named_times_synonyms)?
            if time_string in named_times_combined:
                log.debug("  named time frame detected")
...
```

If you give your okay, I can do that (I think without breaking any functionality) and make a pull request.

**Comment by Daenara** [comment]
subfunctions would be okay but the way you put your example makes the parser a class and that is something I want to avoid. Just handing the request to each subfunction would be better. If you think yo can manage, feel free, otherwise it will have to wait until I find time to do it.

**Comment by ulno** [comment]
Sorry, was just a bit hasty typing the example up, didn't want to enforce any object orientation where it is not necessary. I will see if I have time for some refactoring into simple subfunctions tomorrow. 

Side note:
I just added a json-option to my command line parser, so you can feed the intent-event-object to it too - this will allow to embed this into node-red - this is then directly forwarded to your rhasspy_intent and should work the same way as you initially intended.

**Comment by Daenara** [comment]
Well, initially the whole thing was intended to work with snips (hence the few comments in the code that I didn't remove yet). I am just glad that I used a somewhat object orientated approach back then so I could easily convert it to use a different way to start it.

I saw the json parser part, seems a lot cleaner than my big pile of ifs and elses. The wy it looks it might give you a problem with the time part thought. Because rhasspy has no better way to input time it is "5 29" right now and no one would actually write a time like this. I have hopes for a time and date datatype somewhere in 2.5 but until then it is this workaround.
